server/controllers/finance/reports/income_expense_by_year/index.js

```diff
@@ -0,0 +1,308 @@
+/**
+ * Income Expense by Month Controller
+ *
+ * This controller is responsible for processing the income and expense by month
+ * report.
+ *
+ * @requires lodash
+ * @requires lib/db
+ * @requires lib/ReportManager
+ * @requires lib/errors/BadRequest
+ * @requires lib/Tree
+ * @requires controllers/fiscal
+ */
+
+
+const _ = require('lodash');
+const db = require('../../../../lib/db');
+const ReportManager = require('../../../../lib/ReportManager');
+const Tree = require('../../../../lib/Tree');
+const util = require('../../../../lib/util');
+
+const TEMPLATE = './server/controllers/finance/reports/income_expense_by_year/report.handlebars';
+
+const INCOME_TYPE_ID = 4;
+const EXPENSE_TYPE_ID = 5;
+const TITLE_ID = 6;
+
+const DEFAULT_PARAMS = {
+  csvKey : 'rows',
+  filename : 'TREE.INCOME_EXPENSE_BY_MONTH',
+  orientation : 'landscape',
+  footerRight : '[page] / [toPage]',
+};
+
+// expose to the API
+exports.document = document;
+
+async function document(req, res, next) {
+  let report;
+
+  const options = _.defaults(req.query, DEFAULT_PARAMS);
+
+  const removeUnusedAccounts = !!+options.removeUnusedAccounts;
+
+  try {
+    report = new ReportManager(TEMPLATE, req.session, options);
+  } catch (e) {
+    next(e);
+    return;
+  }
+
+  const { fiscalYearId } = options;
+  const data = {};
+
+
+  const thirdYear = await getFiscalById(fiscalYearId);
```

Comment:
You can actually do all this at the same time for a performance boost:
```js
const [thirdYear, secondYear, firstYear] = await Promise.all([
  getFiscalById(fiscalYearId),
  getFiscalById(thirdYear.previous_fiscal_year_id),
  getFiscalById(secondYear.previous_fiscal_year_id),
]);
```

[Here is a related experiment with Promise.all() to improve performance](http://www.codingpedia.org/ama/parallel-calls-with-async-await-in-javascript-i-promise-you-all-performance-and-simplicity).

Reply:
I don't need to get all those informations at the same time, there are linked, the previous_fiscal_year_id of a fiscalyear  helps to get another fiscal year